File: api/src/main/java/dev/akif/crud/simplest/SimplestMapper.java
package dev.akif.crud.simplest;

import dev.akif.crud.CRUDDTOMapper;
import dev.akif.crud.CRUDMapper;

import java.io.Serializable;

/**
 * The simplest variant of {@link CRUDMapper} where the only data model is the entity,
 * which is also a {@link CRUDDTOMapper}
 *
 * @param <I> ID type of the entity
 * @param <E> Entity type
 */
public interface SimplestMapper<
        I extends Serializable,
        E extends SimplestEntity<I, E>> extends CRUDMapper<I, E, E, E, E>, CRUDDTOMapper<I, E, E, E, E, E, E, E> {
    @Override
    default E modelToDTO(final E model) {
        return model;
    }

    @Override
    default E entityToModel(final E entity) {
        return entity;
    }
}

File: api/src/main/kotlin/dev/akif/crud/simplest/SimplestMapper.kt
package dev.akif.crud.simplest

import dev.akif.crud.CRUDDTOMapper
import dev.akif.crud.CRUDMapper
import dev.akif.crud.common.Parameters
import java.io.Serializable
import java.time.Instant

/**
 * The simplest variant of [CRUDMapper] where the only data model is the entity,
 * which is also a [CRUDDTOMapper]
 *
 * @param I Id type of the data
 * @param E Entity type of the data which is a [SimplestEntity]
 */
interface SimplestMapper<I : Serializable, E : SimplestEntity<I>> : CRUDMapper<I, E, E, E, E>,
    CRUDDTOMapper<I, E, E, E, E, E, E> {
    override fun modelToDTO(model: E, parameters: Parameters): E = model

    override fun entityToModel(entity: E): E = entity

    override fun createDTOToCreateModel(createDTO: E, parameters: Parameters): E = createDTO

    override fun updateDTOToUpdateModel(updateDTO: E, parameters: Parameters): E = updateDTO

    override fun entityToBeCreatedFrom(createModel: E, now: Instant): E = createModel
}
